### ISSUE TITLE
tensorflow 알고맂므

### ISSUE THREAD

**Issue by RakunKo** [RakunKo/iot8#3]
https://github.com/ludlows/CSI-Activity-Recognition?tab=readme-ov-file

**Comment by RakunKo** [comment]
import numpy as np
import tensorflow as tf

def truncate(num, n):
    integer = int(num * (10 ** n)) / (10 ** n)
    return float(integer)


def sniffing(nicname, real_time_model):
    print('Start Sniffing... @', nicname, 'UDP, Port 5500')
    sniffer = pcap.pcap(name=nicname, promisc=True, immediate=True, timeout_ms=50)
    sniffer.setfilter('udp and port 5500')

    before_ts = 0.0

    for ts, pkt in sniffer:
        if int(ts) == int(before_ts):
            cur_ts = truncate(ts, 1)
            bef_ts = truncate(before_ts, 1)

            if cur_ts == bef_ts:
                before_ts = ts
                continue

        eth = dpkt.ethernet.Ethernet(pkt)
        ip = eth.data
        udp = ip.data

        # MAC Address 추출
        # UDP Payload에서 Four Magic Byte (0x11111111) 이후 6 Byte는 추출된 Mac Address 의미
        mac = udp.data[4:10].hex()

        # Four Magic Byte + 6 Byte Mac Address + 2 Byte Sequence Number + 2 Byte Core and Spatial Stream Number + 2 Byte Chanspac + 2 Byte Chip Version 이후 CSI
        # 4 + 6 + 2 + 2 + 2 + 2 = 18 Byte 이후 CSI 데이터
        csi = udp.data[18:]

        bandwidth = ip.__hdr__[2][2]
        nsub = int(bandwidth * 3.2)

        # Convert CSI bytes to numpy array
        csi_np = np.frombuffer(
            csi,
            dtype=np.int16,
            count=nsub * 2
        )

        # Cast numpy 1-d array to matrix
        csi_np = csi_np.reshape((1, nsub * 2))

        # Convert csi into complex numbers
        csi_cmplx = np.fft.fftshift(
            csi_np[:1, ::2] + 1.j * csi_np[:1, 1::2], axes=(1,)
        )
        
        csi_amp = list(np.abs(csi_cmplx)[0])
        preprocessed_data = real_time_model.preprocess(csi_amp)
        
        if preprocessed_data is not None:
            # 실시간 예측
            prediction = real_time_model.predict(preprocessed_data)
            predicted_activity = np.argmax(prediction, axis=1)
            print(f"Predicted Activity: {predicted_activity}")

        before_ts = ts




class RealTimeCSIModel:
    def __init__(self, model_path, win_len=1000, downsample=2):
        self.model = tf.keras.models.load_model(model_path, custom_objects={'AttenLayer': AttenLayer})
        self.win_len = win_len
        self.downsample = downsample
        self.buffer = []

    def preprocess(self, csi_data):
        # 새로운 데이터를 버퍼에 추가하고 윈도우가 충족되면 반환
        self.buffer.extend(csi_data)
        if len(self.buffer) >= self.win_len:
            window = np.array(self.buffer[:self.win_len])
            self.buffer = self.buffer[len(self.buffer) - self.win_len // 2:]  # 슬라이딩 윈도우 적용
            window = window[::self.downsample, :]  # 다운샘플링
            return window[np.newaxis, ...]
        return None

    def predict(self, preprocessed_data):
        return self.model.predict(preprocessed_data)


class AttenLayer(tf.keras.layers.Layer):
    def __init__(self, num_state, **kw):
        super(AttenLayer, self).__init__(**kw)
        self.num_state = num_state

    def build(self, input_shape):
        self.kernel = self.add_weight(name='kernel', shape=[input_shape[-1], self.num_state])
        self.bias = self.add_weight(name='bias', shape=[self.num_state,])
        self.prob_kernel = self.add_weight(name='prob_kernel', shape=[self.num_state])

    def call(self, input_tensor):
        atten_state = tf.tanh(tf.tensordot(input_tensor, self.kernel, axes=1) + self.bias)
        logits = tf.tensordot(atten_state, self.prob_kernel, axes=1)
        prob = tf.nn.softmax(logits)
        weighted_feature = tf.reduce_sum(tf.multiply(input_tensor, tf.expand_dims(prob, -1)), axis=1)
        return weighted_feature

    # for saving the model
    def get_config(self):
        config = super().get_config().copy()
        config.update({
            'num_state': self.num_state,})
        return config
    
def main():
    model = tf.keras.models.load_model('/Users/gohyeoncheol/iot/CSI-Activity-Recognition/best_atten.keras', custom_objects={'AttenLayer': AttenLayer})


    real_time_model = RealTimeCSIModel('/Users/gohyeoncheol/iot/CSI-Activity-Recognition/best_atten.keras')
    print(real_time_model)

    # 모델의 구조 출력
    print(model.summary())

    # 모델의 요약 정보 출력
    print(model)

    # 네트워크 인터페이스 이름 설정
    nicname = 'wlan0'  # 실제 사용 중인 네트워크 인터페이스 이름으로 변경 필요

    # 실시간 데이터 수집 및 예측 시작
    sniffing(nicname, real_time_model)

if __name__ == "__main__":
    main()
